man/man3/io_nemo.3
.TH IO_NEMO 3NEMO "29 May 2007"
.SH NAME
\fBio_nemo\fP \(em High level C function to read/save NEMO snapshot from a
C program.
.SH SYNOPSIS
.nf
.PP
\fBio_nemo(filename, param, ....)\fP
\fBchar * filename;/* input/output file name */\fP
\fBchar * param;   /* parameters list */\fP
.SH DESCRIPTION
\fIio_nemo\fP is a high level C function which allows the user to
perform I/O operations on NEMO snapshot files from a C program. It
works quite like C printf function.
.PP
The use of this function is very simple and require only some basic
knowledge of how to program in NEMO.
.PP
.SH PARAMETERS
 
.TP 10
\fBfilename\fP=\fIcharacter variable\fP
Input/Output file name of the \fINEMO\fP snapshot
that you want to read/save.
.TP
\fBparam\fP=\fIcharacter\fP
Param is a string in which you specify what you want to do with
the NEMO file. Each choice is defined with a flag separated with ','.
There are two kind of flags. (1) \fIInformation flags\fP allow to
specify some actions during the I/O.
(2) \fIVariables flags\fP allow to specify what you want to get/put
from/into the NEMO file. The FLAGS list is described below.
.PP
.SH (1) Information flags
.TP 10
\fBs | save\fP
Specify that you want to save data to a NEMO file.
.TP 
\fBr | read\fP
Specify that you want to read data from a NEMO file.
.TP
\fBfloat|single\fP
Specify that the variables that you use to get/put data from/into NEMO
files have been declared in \fIsingle precision\fP. (float).
.TP
\fBdouble\fP
Specify that the variables you use to get/put data from/into NEMO
files have been declared in \fIdouble precision\fP. (double).
.TP
\fBinfo | diag\fP
Gives some informations during the runtime execution.
.TP
\fBclose\fP
Specify that you want to close the NEMO file.
.PP

.SH (2) Variables flags

.TP 10
\fBn|nbody\fP
Match to the number of \fIbodies\fP.
.TP
\fBt|time\fP
Match to the snapshot \fItime\fP.
.TP
\fBm|mass\fP
Match to the particle \fImasses\fP.
.TP
\fBx|pos\fP
Match to the particle \fIpositions\fP.
.TP
\fBv|vel\fP
Match to the particle \fIvelocities\fP.
.TP
\fBp|pot\fP
Match to the particle \fIpotentials\fP.
.TP
\fBa|acc\fP
Match to the particle \fIaccelerations\fP.
.TP
\fBe|eps\fP
Match to the particle \fIsoftening\fP.
.TP
\fBh|history\fP
Match to the NEMO file name that you want to keep track about the history.
.TP
\fBst\fP
Select snapshot's times. You have to  put the selected
time in a string. Example "100:150"
select the time beween 100 and 150. The string is fully compatible
with the expression usually given to \fBnemoinpi\fP (Cf man nemoinpi).
.TP
\fBsp\fP
Select snapshot's particles. You have to  put the selected
particles in a string. Example "0:299999"
select the particles beween 0 and 299999. The \fBreturned nbody\fP
corresponds to the \fBselected\fP nbody.
In the example above it would thus be 300000. The string is fully compatible
with the expression usually given to \fBnemoinpi\fP (Cf man nemoinpi).
.TP
\fBb\fP
Integer variable to store input/output bit flags (see
$NEMOINC/snapshot/snapshot.h). During \fBread\fP operation, this
variable is filled up with OR bit operation according to the
components requested and the existing components in the file. Example:
if you ask for positons and potentials and no potentials exist on the
snapshot, then bit flags will contain only PosBit. During \fBwrite\fP
operation, if it is used, io_nemo engine will only save components, which have
their bits positionned on it, AND requested to be saved. (see a
complete example here: $NEMOSRC/nbody/io_nemo/test_src/io_nemo_test.c).

.SH ARGUMENTS
The \fIvariable flags\fP specify which data you will
get/put from/into the NEMO file, hence after the selection
string \fIparam\fP, you must insert all the selected
variables (arguments) in the same order with which they have been
declared in \fIparam\fP. For each of the following requested flags:
\fIn,t,m,x,v,a,p,b,k,e\fP you **must** give as argument an address of
pointer. During read opreation, io_nemo check NULL pointer for each
arguments. In case of NULL pointer the memory is automatically
allocated to fill up with the requested component. If not NULL, io_nemo
supposed that the pointer is already allocated. For the flags \fIsp,st,h\fP
you just have to give the pointer as argument.

.SH RETURN VALUE
The function return 1 if successfull, 0 if the end of snapshot has been
reached, or -1 if at least one of the components requested does not exist (in that
case you should use 'b' parameter to filter existing components).

.SH EXAMPLE
Here is a simple NEMO C program to illustrate the use of the function
\fIio_nemo()\fP. The program reads all the time steps into a NEMO a
file, and saves the data in another NEMO file. (See a more complete
example here: $NEMOSRC/nbody/io_nemo/test_src/io_nemo_test.c).
.PP
.nf
/* ----------------------------------------------------- *\\ 
|* NEMO program test using 'io_nemo()' function
\\* ----------------------------------------------------- */
#include <stdinc.h>
#include <getparam.h>

string defv[]=  
{ "in=???\\n       input snapshot",
  "out=???\\n      output snapshot",
  "VERSION=1.0\\n  16-April-97 jcl",
  NULL
};

string usage="NEMO program test using 'io_nemo()' function";
/* ----------------------------------------------------- *\\ 
|* Main program
\\* ----------------------------------------------------- */
nemo_main()
{
  float * pos=NULL, * vel=NULL, * mass=NULL;
  float * time=NULL;
  int   * nbody=NULL;
  string in,out;
  int i,j;

  in  =  getparam("in");
  out =  getparam("out");

  i = 1;
  while (i!=0) {
      /* read the NEMO snapshot */
      i=io_nemo(in,"float,n,t,x,v,m,read,info",
		&nbody,&time,&pos,&vel,&mass);

      /* save the NEMO snapshot */
      if (i != 0)
	j=io_nemo(out,"float,n,t,x,v,m,h,save,info",
		  &nbody,&time,&pos,&vel,&mass,in);

  }
  /* close the output NEMO snapshot */
  io_nemo(out,"close");
}
/* ----------------------------------------------------- *\\ 
|* END
\\* ----------------------------------------------------- */
.SH IMPORTANT THINGS
a) Notice in the example above, that in the parameter list, 'n' matches
\'nbody', 'm' matches 'mass', 'x' matches 'pos', 'v' matches 'vel', 't' matches 'ts'. 
All the variables are in the same order that they have been
declared in the \fIparam list\fP.

b) All the NEMO variables (nbody, time, mass, pos, vel, acc, pot) must
be declared as a \fIpointer\fP equal to \fINULL\fP. During the
io_nemo() call, theses pointers will be allocated according the number
of bodies in the snapshot, so you must give the \fIadress\fP of the pointer
( ie : &pos, &vel, ...etc).

c) All the floating arrays must be declared in the same floating type.

d) During a "read" operation, the function io_nemo() return '0' if
it is the end of the NEMO file. That means that no new values have
been read.

.fi
.PP
.SH COMPILATION
To use the function \fIio_nemo()\fP from a C program you must
link your program with the library \fIlibnemo.a\fP 
as described in the Makefile below.
.PP
.nf
# ----------------------------------------
# MAKEFILE to use IO_NEMO
# 
# ----------------------------------------

# path for NEMO Library and IO_NEMO library
LIBS = -L$(NEMOLIB) -L/usr/local/lib

io_nemo_test : io_nemo_test.o
	$(CC) -o $@ io_nemo_test.o $(LIBS) \\
                  -lnemo -lm

# ----------------------------------------
.fi
.SH SEE ALSO
.nf
nemo(1NEMO), snapshot(5NEMO).
.fi
.SH AUTHOR
Jean-Charles LAMBERT

.SH BUGS and COMMENT
Please, report all bugs and comment to : 
.nf
Jean-Charles.Lambert@oamp.fr
.fi
.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
15-Nov-96	 V1.0 : created                           JCL
21-Feb-97	 V1.1 : memory optimisation               JCL
16-Apr-97	 V1.11: manual created                    JCL 
19-Jul-02	 V1.2 : io_nemo and io_nemo_f unified     JCL
18-Mar-04	 V1.21: bugs fixed, softening added       JCL
03-Mar-05	 V1.30: memory bugs fixed, keybits added  JCL
	                valgrind mem/leak safe            
24-Apr-06	 V1.31: memory leak fixed                 JCL
19-Jun-06	 V1.32: happy gfortran                    JCL
29-May-07	 V1.42: handle snapshot with different    JCL
	                nbodies   

.fi


src/nbody/io_nemo/src/io_nemo.c
/* =================================================================
|  Copyright Jean-Charles LAMBERT - 2006                            
|  e-mail:   Jean-Charles.Lambert@oamp.fr                           
|  address:  Dynamique des galaxies                                 
|            Laboratoire d'Astrophysique de Marseille               
|            2, place Le Verrier                                    
|            13248 Marseille Cedex 4, France                        
|            CNRS U.M.R 6110                                        
| ==================================================================
|* Perform I/O operations on NEMO structure data from a C program   
| ==================================================================
| history                                                           
|                *            *             *                       
| 15-Nov-96	 V1.0 : created                                  JCL
| 21-Feb-97	 V1.1 : memory optimisation                      JCL
| 16-Apr-97	 V1.11: manual created                           JCL
| 19-Jul-02	 V1.2 : io_nemo and io_nemo_f unified            JCL
| 18-Mar-04	 V1.21: bugs fixed, softening added              JCL
| 03-Mar-05	 V1.30: memory bugs fixed, nemo control bits     JCL
|                       added, valgrind mem/leak safe               
| 24-Apr-06      V1.31: memory leak fixed                        JCL
| 19-Jun-06      V1.32: happy gfortran                           JCL
| 29-May-07      V1.42: handle snapshot with different #bodies   JCL
+----------------------------------------------------------------  */

/* -----------------------------------------------------------------
|  Include files                                                    
+----------------------------------------------------------------- */
#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>		
#include <filestruct.h>
#include <history.h>

#include <snapshot/snapshot.h>	
#include <snapshot/body.h>

#include <stdio.h>
#include <stdarg.h>

#include "io_init.h" 
#include "parameters.h"
#include "io_get_put.h"
#include "check_file.h"
#include "io_nemo.h"
#include "maxio.h"

/* extern variables */
#include "flags_data.h"

/* -----------------------------------------------------------------
|  Shared variables                                                 
+----------------------------------------------------------------- */

/* Variables used for reading operations */
char * io_in[MAXIO]; 
FILE * instr[MAXIO];
bool   read_one[MAXIO];

/* Variables used for writing operations */
char * io_out[MAXIO]; 
FILE * outstr[MAXIO];
bool   save_one[MAXIO];

/* variables used to keep track of history (EXPORTED) */
bool   set_history[MAXIO];
char * history_prog=NULL;
char * hist_file;         /* history file name  */

/* variables to store max #bodies per file */
int   maxbodies[MAXIO];

/* ------------------------------------------------------------------
|  reajust_ptr :                                                     
|  match for each pointers passing to "io_nemo" the pointers         
|  allocated.                                                        
+------------------------------------------------------------------ */ 
void reajust_ptr()
{
#if 0 
  if (N_io)    *nbody_p    = nbody;
  if (T_io)    *time_p     = timu;
  if (M_io)    *mass_p     = mass;
  if (X_io)    *pos_p      = pos;
  if (V_io)    *vel_p      = vel;
  if (XV_io)   *phase_p    = phase;
  if (P_io)    *pot_p      = pot;
  if (A_io)    *acc_p      = acc; 
  if (K_io)    *keys_p     = keys; 
  if (EPS_io)  *eps_p      = eps; 
  /*  if (ST_io)   *selt_p     = selt; */
#endif
}

/* ------------------------------------------------------------------
|  io_nemo :                                                         
|  function called from a C program to perform I/O on a NEMO file.   
+------------------------------------------------------------------ */
int io_nemo(char * iofile,
            char * param, ...)
{
  va_list pa;      /* variable parameter list                 */


  t_ion_data * ion;

  char * p, * field;
  int code;        /* return value according to I/O performed */
  bool io_op=TRUE; /* TRUE -> READ,   FALSE -> SAVE           */
  int rtype;       /* rtype = 1 (float), or = 2 (double)      */
  /* control */
  static bool first=TRUE;


  /* init flag */
  init_flag_io();

  /* init I/O variable */
  if (first) {
    init_io_one(maxbodies, read_one, save_one, set_history, &history_prog, MAXIO);
    first = FALSE;
  }

  /* allocate memory for a new data structure */
  ion = (t_ion_data *) malloc(sizeof(t_ion_data));
  if (!ion) {
    fprintf(stderr,"Unable to allocate memory of size [t_ion_data], aborting...\n");
    exit(1);
  }

  /* get the first parameter from the variable list */
  va_start(pa, param);
  p = param;    /* point to the first element of the variable list */

  while (*p) {

    field = (char *) get_field(&p); /* get the next element */
    switch (get_case(field)) {

    case 1 : N_io = 1;
      ion->nbody_p = va_arg(pa, int **);
      ion->nbody   = *(ion->nbody_p); /* match local and parameter pointer */
      break;

    case 2 : T_io = 1;
      ion->time_p  = va_arg(pa, char **);
      ion->timu    = *(ion->time_p); /* match local and parameter pointer */
      break;

    case 3 : M_io = 1;
      ion->mass_p  = va_arg(pa, char **);
      ion->mass    = *(ion->mass_p); /* match local and parameter pointer */
      break;

    case 4 : X_io = 1;
      ion->pos_p = va_arg(pa, char **);
      ion->pos   = *(ion->pos_p);
      break;

    case 5 : V_io = 1;
      ion->vel_p   = va_arg(pa, char **);
      ion->vel     = *(ion->vel_p); /* match local and parameter pointer */
      break; 

    case 6 : P_io = 1;
      ion->pot_p   = va_arg(pa, char **);
      ion->pot     = *(ion->pot_p); /* match local and parameter pointer */
      break;

    case 7 : A_io = 1;
      ion->acc_p   = va_arg(pa, char **);
      ion->acc     = *(ion->acc_p); /* match local and parameter pointer */
      break;

    case 8 : K_io = 1;
      ion->keys_p   = va_arg(pa, char **);
      ion->keys     = *(ion->keys_p); /* match local and parameter pointer */
      break;

    case 10 : EPS_io = 1;
      ion->eps_p   = va_arg(pa, char **);
      ion->eps     = *(ion->eps_p); /* match local and parameter pointer */
      break;

    case 11 : B_io = 1;
      ion->bits_p   = va_arg(pa, int **);
      ion->bits     = *(ion->bits_p); /* match local and parameter pointer */
      break;

    case 57 : ST_io = 1;
      ion->selt     = va_arg(pa, char *);
      break;

    case 52 : io_op = FALSE;  /* save the snapshot */
      break;

    case 53 : io_op = TRUE;   /* read the snapshot */
      break;

    case 54 : rtype = 1;     /* float dimension */
      break;

    case 55 : rtype = 2;     /* double dimension */
      break;

    case 56 : I_io = 1;      /* print information */
      break;

    case 59 : H_io = 1;      /* History file name */
      hist_file = va_arg(pa, char *);
      break;
	  
    case 58 : SP_io=1;
      ion->SelectionString123=va_arg(pa, char*);
      break;

    case 60 : C_io = 1;
      break;

    default  : fprintf(stderr,
		       "Parameter error ## [io_nemo] \"%s\" unknown\n",
		       field);
    exit(1);
    break;
    }

    free(field);
  }

  va_end(pa);

  if (C_io)  /* close the snapshot */
    code=close_io_nemo(iofile);
  else {
    if (io_op) { /* operation on snaphot is reading */
      code = get_data_select(iofile, rtype, io_in, read_one, instr, MAXIO,ion);
      /* reajust pointers */
      /*reajust_ptr();*/
      if (N_io)    *(ion->nbody_p)    = ion->nbody;
      if (T_io)    *(ion->time_p )    = ion->timu;
      if (M_io)    *(ion->mass_p )    = ion->mass;
      if (X_io)    *(ion->pos_p  )    = ion->pos;
      if (V_io)    *(ion->vel_p  )    = ion->vel;
      if (XV_io)   *(ion->phase_p)    = ion->phase;
      if (P_io)    *(ion->pot_p  )    = ion->pot;
      if (A_io)    *(ion->acc_p  )    = ion->acc; 
      if (K_io)    *(ion->keys_p )    = ion->keys; 
      if (EPS_io)  *(ion->eps_p  )    = ion->eps; 
      if (B_io)    *(ion->bits_p )    = ion->bits; 
    }
    else { 
      if (!N_io) {
	fprintf(stderr,"Parameter error ## [io_nemo] param : \"%s\"\n",
		param);
	fprintf(stderr,
		"You must specify \"nbody\" "
		"in the field parameter for SAVE operation \n");
	exit(1);
      }
      code = put_data_select(iofile, rtype, io_out, save_one, outstr, MAXIO,ion);
    }
  }
  free ((t_ion_data *) ion);
  return code;
}
/* ------------------------------------------------------------------
|  close_io_nemo :                                                   
|  Close the opening snapshot.                                       
|  Return 0 if the file was not open, otherwise 1                    
+------------------------------------------------------------------ */
int close_io_nemo(char * iofile)
{
  int no_io,code;
  
  /* check if the file is already open */
  if ((no_io = get_old_file(iofile,io_in,read_one,instr,MAXIO)) < 0) {
    if ((no_io = get_old_file(iofile,io_out,save_one,outstr,MAXIO)) < 0) {
      fprintf(stderr,
	      "WARNING!! snapshot [%s] not OPEN, unable to close it\n",
	      iofile);
      code=0;
    } 
    else {
      /* close the file open for writing */
      strclose(outstr[no_io]);

      /* RAZ variables */
      save_one[no_io] = FALSE;
      set_history[no_io] = FALSE;
	  
      free((char *) io_out[no_io]);
      code=1;
    }
  }
  else {
    /* close the file open for reading */
    strclose(instr[no_io]);

    /* RAZ variables */
    read_one[no_io] = FALSE;
    maxbodies[no_io]= 0;
    set_history[no_io] = FALSE;
    free((char *) io_in[no_io]);
    code=1;
  }
  /* reset_history();  */
  return code;
}
/* ------------------------------------------------------------------
|  End of io_nemo.c                                                  
+------------------------------------------------------------------ */
